rfcs/proposed/task_group_dynamic_dependencies/implementation_details.md

```diff
@@ -0,0 +1,724 @@
+`function_stack_task` is created only when `task_group::run_and_wait` is used.
+
+The difference between these types of tasks is that `function_task` copies the body
+to itself and `function_stack_task` uses a reference to the use body instance since
+it can't be destroyed before the blocking API finishes.
+
+Due to API limitations, only the `function_task` can be owned by the `task_handle`.
+
+`function_task` is inherited from the `task_handle_task` object that manages the lifetime of the task.
+`task_handle_task` and `function_stack_task` are inherited from the basic `task` class.
+
+`task_handle` class is implemented as a `unique_ptr` owning the `task_handle_task` instance.
+
+The class layout is shown on the picture below:
+
+<img src="assets/impl_prev_class_layout.png" width=600>
+
+### `task_dynamic_state` class
+
+The main part of the APIs described above are implemented as part of the new
+`task_dynamic_state` class. It is intended to manage the task status (in progress, completed, transferred),
+current list of successors and linkage with the new task after the transferring.
+Its layout would be described in details in the later sections. 
+
+Each task in the task group that can have predecessors or successors has the `task_dynamic_state` instance associated with it.
+
+Since `make_edge` API allows completed tasks as predecessors, it is required to prolong the lifetime of the `task_dynamic_state`
+instance associated with the task even after the task completion until the last `tbb::task_tracker` object associated with the
+task would be destroyed.
+
+### `task_with_dynamic_state` class
+
+To associate the instance of `task_dynamic_state` with the task, the current implementation adds the new class `task_with_dynamic_state`
+in the schema above between `task_handle_task` and `task` classes. This class contains the atomic pointer to the `task_dynamic_state`
+instance associated with the task:
+
+<img src="assets/impl_new_class_layout.png" width=1000>
+
+```cpp
+namespace oneapi {
+namespace tbb {
+
+class task_dynamic_state { /*...*/ };
+
+class task { /*...*/ };
+
+class task_with_dynamic_state : public task {
+public:
+    task_with_dynamic_state();
+    ~task_with_dynamic_state();
+
+    task_dynamic_state* get_dynamic_state();
+    void complete_task();
+private:
+    std::atomic<task_dynamic_state*> m_state;
+};
+
+class task_handle_task : public task_with_dynamic_state { /*...*/ };
+class function_task : public task_handle_task { /*...*/ };
+
+} // namespace tbb
+} // namespace oneapi
+```
+
+The main responsibility of the `task_with_dynamic_state` class is to lazily create the `task_dynamic_state` object associated
+with the task and provide access to the created object.
+
+Default constructor of the `task_with_dynamic_state` does not create a `task_dynamic_state` instance associated with the task
+not to introduce additional overhead on creating the instance if the dynamic task dependencies feature is not used.
```

Comment:
```suggestion
To avoid introducing possibly unnecessary overhead, the default constructor of the `task_with_dynamic_state`
does not create a `task_dynamic_state` instance.
```